sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -527,13 +528,141 @@ def test_agg_n_clusters():
-               " %s was provided." % str(agc.n_clusters))
-        assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = -1
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("n_clusters should be an integer greater than 0."
+           " %s was provided." % str(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+    n_clus = 0
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
+           "distance_threshold=None instead.".format(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+
+
+def test_agg_n_cluster_or_distance_threshold():
+    # Test that an error is raised when n_clusters
+    # and distance_threshold are None
+
+    n_clus = None
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    # distance_threshold is None by default
+    agc = AgglomerativeClustering(n_clusters=n_clus)
+    msg = ("Either n_clusters (>0) or distance_threshold "
+           "needs to be set, got n_clusters={} and "
+           "distance_threshold=None instead.".format(agc.n_clusters))
+    assert_raise_message(ValueError, msg, agc.fit, X)
+
+
+def test_agg_n_cluster_and_distance_threshold():
+    # Test that when distance_threshold is set
+    # n_clusters_ is set to None
+
+    n_clus, dist_thresh = 10, 10
+    rng = np.random.RandomState(0)
+    X = rng.rand(20, 10)
+    agc = AgglomerativeClustering(n_clusters=n_clus,
+                                  distance_threshold=dist_thresh)
+    agc.fit(X)
+    assert_true(agc.n_clusters == n_clus)
+    assert_true(agc.n_clusters_ is None)
 
 
+def test_agglomerative_clustering_with_distance_threshold():
+    # Check that we obtain the correct number of clusters with
+    # agglomerative clustering with distance_threshold.
+
+    rng = np.random.RandomState(0)
+    mask = np.ones([10, 10], dtype=np.bool)
+    n_samples = 100
+    X = rng.randn(n_samples, 50)
+    connectivity = grid_to_graph(*mask.shape)
+    # test when distance threshold is set to 10
+    distance_threshold = 10
+    for linkage in ("ward", "complete", "average"):
+        clustering = AgglomerativeClustering(
+            n_clusters=None,
+            distance_threshold=distance_threshold,
+            connectivity=connectivity, linkage=linkage)
+        clustering.fit(X)
+        clusters_produced = clustering.labels_
+        num_clusters_produced = len(np.unique(clustering.labels_))
+        # test if the clusters produced match the point in the linkage tree
+        # where the distance exceeds the threshold
+        tree_builder = _TREE_BUILDERS[linkage]
+        children, n_components, n_leaves, parent, distances = \
+            tree_builder(X, connectivity=connectivity, n_clusters=None,
+                         return_distance=True)
+        clusters_at_threshold = _hc_cut(n_clusters=None, children=children,
```

Comment:
please move this down to "test clusters produced". I also think this should be using the `n_clusters` parameter rather than the `distance_threshold` parameter, to effectively test equivalence.